src/cluster_legacy.c

```diff
@@ -1671,17 +1671,40 @@ clusterNode *createClusterNode(char *nodename, int flags) {
     node->tcp_port = 0;
     node->cport = 0;
     node->tls_port = 0;
-    node->fail_reports = listCreate();
+    node->fail_reports = raxNew();
     node->orphaned_time = 0;
     node->repl_offset = 0;
-    listSetFreeMethod(node->fail_reports, zfree);
     node->is_node_healthy = 0;
     return node;
 }
 
+/* 8 bytes mstime + 8 bytes cluster node */
+#define FAILURE_REPORT_ST_KEYLEN 16
+
+/* The expiry time is rounded up (ceiled) to the nearest second in order to bucket
+ * timestamps and keep the radix‐tree keys compact.  The resulting timestamp is stored
+ * in big‑endian format, followed by the pointer to the clusterNode. */
+static void encodeFailureReportKey(unsigned char *buf, mstime_t expiry, clusterNode *node) {
+    mstime_t bt = ((expiry + 999) / 1000) * 1000;
+    uint64_t be = htonu64((uint64_t)bt);
+    memcpy(buf, &be, 8);
+    memcpy(buf + 8, &node, sizeof(node));
+    if (sizeof(node) == 4) memset(buf + 12, 0, 4);
+}
```

Comment:
Let's not store the pointer address. It's error prone, what if the node is already freed (via CLUSTER FORGET) while trying to access it back, this will cause a use-after-free crash. Let's use sds api to form the key  `ts + _ + nodename`.